sdks/python/apache_beam/ml/inference/tensorrt_inference.py

```diff
@@ -0,0 +1,213 @@
+      self.onnx_path = kwargs.get('onnx_path')
+
+    trt.init_libnvinfer_plugins(LOGGER, namespace="")
+
+  def batch_elements_kwargs(self):
+    """Sets min_batch_size and max_batch_size of a TensorRT engine."""
+    return {
+        'min_batch_size': self.min_batch_size,
+        'max_batch_size': self.max_batch_size
+    }
+
+  def load_model(self) -> trt.ICudaEngine:
+    """Loads and initializes a TensorRT engine for processing."""
+    return _load_engine(self.engine_path)
+
+  def load_onnx(self) -> trt.INetworkDefinition:
+    """Loads and parses an onnx model for processing."""
+    return _load_onnx(self.onnx_path)
+
+  #
+  def build_engine(self, network: trt.INetworkDefinition) -> trt.ICudaEngine:
+    """Build an engine according to parsed/created network."""
+    return _build_engine(network)
+
+  def run_inference(
+      self,
+      batch: np.ndarray,
+      engine: trt.ICudaEngine,
+      inference_args: Optional[Dict[str, Any]] = None
+  ) -> Iterable[PredictionResult]:
+    """
+    Runs inferences on a batch of Tensors and returns an Iterable of
+    TensorRT Predictions.
+
+    Args:
+      batch: A np.ndarray or a np.ndarray that represents a concatenation
+        of multiple arrays as a batch.
+      engine: A TensorRT engine.
+      inference_args: Any additional arguments for an inference
+        that are not applicable to TensorRT.
+
+    Returns:
+      An Iterable of type PredictionResult.
+    """
+    _validate_inference_args(inference_args)
+    context = engine.create_execution_context()
+    assert context
+    # Setup I/O bindings
+    inputs = []
+    outputs = []
+    allocations = []
+    for i in range(engine.num_bindings):
+      is_input = False
+      if engine.binding_is_input(i):
+        is_input = True
+      name = engine.get_binding_name(i)
+      dtype = engine.get_binding_dtype(i)
+      shape = engine.get_binding_shape(i)
+      if is_input:
+        batch_size = shape[0]
+      size = np.dtype(trt.nptype(dtype)).itemsize
+      for s in shape:
+        size *= s
+      allocation = cuda.mem_alloc(size)
+      binding = {
+          'index': i,
+          'name': name,
+          'dtype': np.dtype(trt.nptype(dtype)),
+          'shape': list(shape),
+          'allocation': allocation,
+      }
+      allocations.append(allocation)
+      if engine.binding_is_input(i):
+        inputs.append(binding)
+      else:
+        outputs.append(binding)
+
+    assert batch_size > 0
+    assert len(inputs) > 0
+    assert len(outputs) > 0
+    assert len(allocations) > 0
+    # Prepare the output data
+    predictions = []
+    for output in outputs:
+      predictions.append(np.zeros(output['shape'], output['dtype']))
+    # Process I/O and execute the network
+    cuda.memcpy_htod(inputs[0]['allocation'], np.ascontiguousarray(batch))
+    context.execute_v2(allocations)
+    for output in range(len(predictions)):
+      cuda.memcpy_dtoh(predictions[output], outputs[output]['allocation'])
```

Comment:
You might see some perf benefit by overlapping execution/memcpy's by using `memcpy_htod_async`, `execute_async_v2`, and `memcpy_dtoh_async`.

Reply:
Noted